File: frontend/src/types/autonomous.ts
export interface ActionDefinition {
  action_type: string;
  domain: string;
  provider: string;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  description: string;
  required_permissions: string[];
  supports_dry_run: boolean;
  supports_simulation: boolean;
  supports_rollback: boolean;
  requires_approval: boolean;
}

export interface RemediationPlan {
  id: string;
  trigger_source: string;
  source_event_id?: string;
  root_cause: string;
  affected_resource: string;
  provider: string;
  environment: string;
  action_type: string;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  expected_impact: string;
  estimated_downtime_sec: number;
  estimated_cost_impact: number;
  requires_approval: boolean;
  rollback_supported: boolean;
  execution_mode: 'DRY_RUN' | 'SIMULATED' | 'LIVE';
  confidence_score: number;
  status: 'PLANNED' | 'WAITING_APPROVAL' | 'APPROVED' | 'EXECUTING' | 'VERIFYING' | 'COMPLETED' | 'FAILED' | 'ROLLED_BACK' | 'BLOCKED';
  plan_details: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface RemediationExecution {
  id: string;
  plan_id: string;
  idempotency_key: string;
  execution_mode: 'DRY_RUN' | 'SIMULATED' | 'LIVE';
  status: 'QUEUED' | 'VALIDATING' | 'WAITING_APPROVAL' | 'EXECUTING' | 'VERIFYING' | 'COMPLETED' | 'FAILED' | 'ROLLED_BACK' | 'BLOCKED';
  started_at?: string;
  completed_at?: string;
  precondition_result: Record<string, any>;
  execution_result: Record<string, any>;
  verification_result: Record<string, any>;
  previous_state?: Record<string, any>;
  new_state?: Record<string, any>;
  error_message?: string;
  rollback_status: string;
  created_at: string;
}

export interface AutonomyPolicy {
  id: string;
  autonomy_level: number; // 0-4
  max_autonomous_risk: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  allowed_providers: string[];
  allowed_environments: string[];
  excluded_resources: string[];
  excluded_namespaces: string[];
  default_execution_mode: 'DRY_RUN' | 'SIMULATED' | 'LIVE';
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface RemediationAuditLog {
  id: string;
  plan_id?: string;
  execution_id?: string;
  actor_id?: string;
  action_type: string;
  event_type: string;
  target_resource: string;
  provider: string;
  execution_mode: string;
  details: Record<string, any>;
  created_at: string;
}

export interface AutonomousOverview {
  autonomy_level: number;
  execution_mode: string;
  active_remediations_count: number;
  total_plans_count: number;
  completed_remediations_count: number;
  success_rate_pct: number;
  verification_success_rate_pct: number;
  rollback_rate_pct: number;
  blocked_actions_count: number;
  incidents_prevented_est: number;
  mode_indicator: string;
}

export interface SimulationResult {
  action_type: string;
  affected_resource: string;
  provider: string;
  environment: string;
  execution_mode: string;
  preconditions: Record<string, any>;
  approval_requirement: Record<string, any>;
  simulated_verification: Record<string, any>;
  simulation_result: string;
  message: string;
}

File: backend/app/schemas/autonomous.py
"""
Pydantic Schemas for Autonomous Operations & Self-Healing Center.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

# ── Action Catalog Schemas ───────────────────────────────────────────────────


class ActionDefinitionResponse(BaseModel):
    action_type: str
    domain: str
    provider: str
    risk_level: str
    description: str
    required_permissions: list[str]
    supports_dry_run: bool
    supports_simulation: bool
    supports_rollback: bool
    requires_approval: bool


# ── Remediation Plan Schemas ─────────────────────────────────────────────────


class RemediationPlanCreate(BaseModel):
    trigger_source: str = Field(default="incident_intelligence")
    source_event_id: str | None = None
    root_cause: str
    affected_resource: str
    provider: str = Field(default="AWS")
    environment: str = Field(default="production")
    action_type: str
    risk_level: str = Field(default="MEDIUM")
    expected_impact: str
    estimated_downtime_sec: int = 0
    estimated_cost_impact: float = 0.0
    execution_mode: str = Field(default="SIMULATED")


class RemediationPlanResponse(BaseModel):
    id: uuid.UUID
    trigger_source: str
    source_event_id: str | None
    root_cause: str
    affected_resource: str
    provider: str
    environment: str
    action_type: str
    risk_level: str
    expected_impact: str
    estimated_downtime_sec: int
    estimated_cost_impact: float
    requires_approval: bool
    rollback_supported: bool
    execution_mode: str
    confidence_score: float
    status: str
    plan_details: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ── Execution Schemas ─────────────────────────────────────────────────────────


class RemediationExecutionResponse(BaseModel):
    id: uuid.UUID
    plan_id: uuid.UUID
    idempotency_key: str
    execution_mode: str
    status: str
    started_at: datetime | None
    completed_at: datetime | None
    precondition_result: dict[str, Any]
    execution_result: dict[str, Any]
    verification_result: dict[str, Any]
    previous_state: dict[str, Any] | None
    new_state: dict[str, Any] | None
    error_message: str | None
    rollback_status: str
    created_at: datetime

    class Config:
        from_attributes = True


# ── Policy & Maintenance Window Schemas ───────────────────────────────────────


class AutonomyPolicyUpdate(BaseModel):
    autonomy_level: int = Field(ge=0, le=4, default=1)
    max_autonomous_risk: str = Field(default="LOW")
    allowed_providers: list[str] = Field(default_factory=lambda: ["AWS", "Azure", "GCP", "Kubernetes"])
    allowed_environments: list[str] = Field(default_factory=lambda: ["development", "staging", "production"])
    excluded_resources: list[str] = Field(default_factory=list)
    excluded_namespaces: list[str] = Field(default_factory=list)
    default_execution_mode: str = Field(default="SIMULATED")
    is_active: bool = True


class AutonomyPolicyResponse(BaseModel):
    id: uuid.UUID
    autonomy_level: int
    max_autonomous_risk: str
    allowed_providers: list[str]
    allowed_environments: list[str]
    excluded_resources: list[str]
    excluded_namespaces: list[str]
    default_execution_mode: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MaintenanceWindowCreate(BaseModel):
    title: str
    environment: str = Field(default="production")
    start_time: datetime
    end_time: datetime
    timezone: str = Field(default="UTC")
    block_all_actions: bool = True
    allowed_actions: list[str] = Field(default_factory=list)


class MaintenanceWindowResponse(BaseModel):
    id: uuid.UUID
    title: str
    environment: str
    start_time: datetime
    end_time: datetime
    timezone: str
    block_all_actions: bool
    allowed_actions: list[str]
    created_at: datetime

    class Config:
        from_attributes = True


# ── Simulation & Audit Schemas ────────────────────────────────────────────────


class SimulationRequest(BaseModel):
    action_type: str
    affected_resource: str
    provider: str = Field(default="AWS")
    environment: str = Field(default="production")
    execution_mode: str = Field(default="SIMULATED")


class AuditLogResponse(BaseModel):
    id: uuid.UUID
    plan_id: uuid.UUID | None
    execution_id: uuid.UUID | None
    actor_id: uuid.UUID | None
    action_type: str
    event_type: str
    target_resource: str
    provider: str
    execution_mode: str
    details: dict[str, Any]
    created_at: datetime

    class Config:
        from_attributes = True


class AutonomousOverviewResponse(BaseModel):
    autonomy_level: int
    execution_mode: str
    active_remediations_count: int
    total_plans_count: int
    completed_remediations_count: int
    success_rate_pct: float
    verification_success_rate_pct: float
    rollback_rate_pct: float
    blocked_actions_count: int
    incidents_prevented_est: int
    mode_indicator: str
